samples: mesh: nrf52: to solve pending (PTS) issues after merging PR:9475, support to push device in factory reset mode & other minor improvements

vikrant8052 opened the pull request:
This PR will try to solve issues which are pending after merging https://github.com/zephyrproject-rtos/zephyr/pull/9475

Separately defined publisher functions for Servers. This would help
to send extra status responses whenever required.

This PR adds facility to push device into factory reset mode
using power reset. As per current implementation user have to
do 5 times reset so that gap between two consecutive reset
should not exceed 7 seconds.

With this PR it is now possible to save following states/values on SoC Flash:
- Generic Default Transition Time state
- Generic OnPowerUp state
- Lightness & Temperature Default Values
- Lightness & Temperature Last Values

And other minor improvements.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8052 committed:
samples: mesh: nrf52: added separate Server's Publisher functions

Separately defined functions for Servers. This would help
to send extra status responses whenever required.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
Kindly don't merge as of now.

vikrant8051 commented:
@carlescufi 
could you helped me to resolved 
https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/22312/7/tests
this issue ?

vikrant8051 commented:
- [ ] MMDL/SR/LLN/BV-06-C https://hastebin.com/mediwoduga.sql
- [ ] MMDL/SR/LLN/BV-07-C https://hastebin.com/ajukokoqin.sql
- [ ] MMDL/SR/LLN/BV-08-C https://hastebin.com/ocayekecul.sql
- [ ] MMDL/SR/LLN/BV-09-C https://hastebin.com/nekedobewu.sql
- [ ] MMDL/SR/LLN/BV-10-C https://hastebin.com/afomogiqox.sql
- [ ] MMDL/SR/LLN/BV-11-C https://hastebin.com/uqewahunuk.sql
- [ ] MMDL/SR/LLN/BV-12-C https://hastebin.com/adexujixak.sql
- [ ] MMDL/SR/LLN/BV-13-C https://hastebin.com/nuxapejuqa.sql
- [ ] MMDL/SR/LCTL/BV-04-C https://hastebin.com/ufagubazim.sql

vikrant8052 committed:
samples: mesh: nrf52: support for extra status responses

With this commit, IUT will send additional status responses
whenever required as per PTS requirements.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -114,6 +115,35 @@ static void gen_onoff_get(struct bt_mesh_model *model,
+	}
+}
+
+void gen_onoff_publisher(struct bt_mesh_model *model)
```
After 2nd commit, mentioned issue has solved. This may satisfy PTS requirements (as per above mentioned 9 failed logs). Need PTS to re-run against it to reach at final conclusion.

vikrant8051 replied:
done !!

vikrant8051 commented:
As per 6.4.1.2.3 of Mesh Model Specification,

> When a Light Lightness Actual state change is completed as a result of processing an acknowledged
> message or a transaction of messages other than a Light Lightness Set message, a corresponding
> status message shall be sent to the originator to indicate the status change, and a separate Light
> Lightness Status message shall be published. It is recommended to transmit a Light Lightness Status
> message when the element has been physically turned on or off locally (as opposed to via the mesh
> network).

& as per 6.4.3.2.3

> When a Light CTL state change is completed as a result of processing
> an acknowledged message or a transaction of messages other than a Light CTL Set message, a
> corresponding status acknowledge message shall be sent to the originator and a separate Light CTL
> Status message shall be published.

2nd commit would take care of these two points.

vikrant8051 commented:
Along with review it also needs feedback from PTS.
@MariuszSkamra if time allow you then please run PTS against it.

vikrant8051 commented:
@jhedberg @carlescufi @MariuszSkamra @sjanc
As of now, I've not test performance of this App against Mobile App Seekbar/Scrollbar GUI ( .... while controlling GROUP of Nodes over single GATT Proxy connection).

As per my understanding current implementation could be processor hungry & it will take some time to process the data over mesh when user scroll Seekbar to vary **Gen. Level** or **Light Lightness Actual** or **Light CTL** states. 

If user scroll Seekbar very fast then it could also hang Mesh n/w for while.

1. Solution is there should be only 10 virtual discrete points on Seekbar which are hidden for user.

2. I simply take reference from Bluetooth Mesh Study Guide to call Zephyr APIs
which are related to timer & workqueue Threads. So I need your help/suggestions to improve/optimize this App, to handle this situation so that whenever user vary Seekbar it should look like all nodes are in Sync.

Thank You !!

vikrant8052 committed:
samples: mesh: nrf52: short time multireset bt-mesh unprovisioning

In general, Bluetooth Mesh provisioning process could be failed
because of any reason. Using some h/w interrupt we can call
bt_mesh_reset() or something which will push device into factory
reset mode.

But with some actual products it is not always possible or feasible
to add extra h/w to achieve this goal.

This commit adds facility to push device into factory reset mode
using power reset. As per current implementation user have to
do 5 times reset so that gap between two consecutive reset
should not exceed 7 seconds.

Co-authored-by: Laczen JMS <laczenjms@gmail.com>

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/CMakeLists.txt:
```diff
@@ -19,4 +20,5 @@ zephyr_include_directories(
 	src/
 	src/mesh
 	$ENV{ZEPHYR_BASE}/subsys/bluetooth/controller/hal
+	$ENV{ZEPHYR_BASE}/subsys/bluetooth/host
```
@vikrant8051 your application exists to serve as reference code for developers. I don't think it's a good idea to encourage people to include internal headers and use internal APIs. If you want to hack around feel free to do it for a private project, but I don't think we should be putting such things in sample apps that are part of the upstream tree.

The line above related to controller/hal - what's that for?

vikrant8051 replied:
Sorry, I was not aware about it. But there is no proper reference (for how to settings layer API in App), so I've to do it assuming one of the reviewer will correct me :)

Ok... if I removed $ENV{ZEPHYR_BASE}/subsys/bluetooth/host from CMakeLists.txt & #include "settings.h" from storage.c I'm getting following error:



../src/storage.c:16:11: warning: implicit declaration of function 'BT_SETTINGS_SIZE'; did you mean 'BT_BUF_RX_SIZE'? [-Wimplicit-function-declaration]
  char buf[BT_SETTINGS_SIZE(sizeof(reset_counter))];
           ^~~~~~~~~~~~~~~~
           BT_BUF_RX_SIZE
In file included from /home/vikrant/zephyr/include/misc/slist.h:20:0,
                 from /home/vikrant/zephyr/include/settings/settings.h:12,
                 from ../src/mesh/ble_mesh.h:11,
                 from ../src/storage.c:8:
../src/storage.c: At top level:
../src/storage.c:58:32: error: expected ')' before '(' token
 BT_SETTINGS_DEFINE(ps, ps_set, NULL, NULL);
                                ^
../src/storage.c:42:12: warning: 'ps_set' defined but not used [-Wunused-function]
 static int ps_set(int argc, char **argv, char *val)
            ^~~~~~
Kindly guide me to improve it. Thank You !!

jhedberg replied:
What API did you think I'm talking about when I said "internal API"? I'm talking about BT_SETTINGS. If something is not exposed through a header file in include/bluetooth it's an internal API which is not designed to be used by applications directly.

What's preventing you from choosing some other sub-path than "bt/" and using the settings API directly instead of BT_SETTINGS (which is just a wrapper on top of settings for managing stuff under "bt/")?

jhedberg replied:
@vikrant8051 you could look at subsys/bluetooth/host/settings.c and see how it uses the settings API, and then do something similar (probably simpler) for your app where you use some other sub-path. The reason BT_SETTINGS is not public is that the Bluetooth stack "reserves" for itself everything under the "bt/" path, i.e. it needs to be able to trust that there's nothing else there that can cause conflicts, both now and in future versions of the stack.

vikrant8051 replied:
Ok

vikrant8051 replied:
@Laczen 
Hey under documentation there is now details about what I'm looking for
https://docs.zephyrproject.org/latest/subsystems/settings/settings.html

Your mailing list answer, actually opens new window for me to go one step further.
So it was great help from your side, since first time I could successfully implement 
what I want using settings_layer. :) :+1:

jhedberg replied:
> I think the settings_load needs to be called only in the main to allow other subsystems to use the settings subsystem.

@Laczen the only requirement from the Bluetooth side is that settings_load() is called after the stack has completed initialization. You can do it from main() but then you should either call bt_enable() synchronously (NULL callback) or have some other way to block in main() while waiting for bt_ready() to be called.

vikrant8051 commented:
@Laczen Could you please review or help me to improve the 3rd commit ?

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/CMakeLists.txt:
```diff
@@ -19,4 +20,5 @@ zephyr_include_directories(
 	src/
 	src/mesh
 	$ENV{ZEPHYR_BASE}/subsys/bluetooth/controller/hal
```
I'm still wondering about this line. I realize it was already there before this PR, but could you clarify what it's needed for?

vikrant8051 replied:
I've not added it. (..to be safe I don't remember it or don't know how it get included).

I will remove it & I can compile it without this line.

vikrant8051 commented:
@jhedberg Please review 3rd commit now. 

**CC:** @Laczen

vikrant8052 committed:
samples: mesh: nrf52: storing of some Prime states on SoC flash

With this commit it is now possible to save some Mesh Model
Servers State on SoC flash. This impacts initialization &
state binding after NODE reset & that is why some additional
modification has done as per requirements & removed some
code & comments which are redundant as per current
implementation.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
@sjanc @carlescufi 
Please review this PR. It will be great if you help me to optimize overall App to improve it performance.

vikrant8051 commented:
Any update on PTS front ?

vikrant8051 commented:
@MariuszSkamra @sjanc 
Please execute PTS against this PR. This will help to stabilize the implementation or architecture.
Thanks in advance !!

MariuszSkamra commented:
I'm on sick leave now, coming back on wednesday, hopefully

vikrant8052 committed:
samples: mesh: nrf52: improvements in gen. delta message handler

Now in case of Generic Delta handler, if message with same TID
received after 6 seconds with same source address would consider
as new transaction instead of ignoring it.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8052 committed:
samples: mesh: nrf52: coding style improvements

Removed unwanted lines of codes. In transition.c
define Timers just before they get utilised.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8052 committed:
samples: mesh: nrf52: Vendor Model opcodes redefine

Redefine opcode values for Vendor Model to mimic like
Generic OnOff Server opcode values for sake of convenience.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
@MariuszSkamra Please wait & don't execute PTS against this PR as of now. Thank You !!

vikrant8051 commented:
As per current implementation, if client publishes Gen. Level Value as 100 %
& Transition Time as 60 Secs then in Gen. Level Status response Server provides correct present, target values as well as remaining time to achieve the target value.

But problem is, if in meantime (during 60 secs) client ask for Light CTL status or Light Lightness status, then it would **only** respond with **present** Light CTL / Light Lightness value. Here status would not contain target value & Remaining time. Currently I don't know how to crack it plus I don't know whether PTS is expecting this or not.

@MariuszSkamra So please go ahead as of now on PTS front.

vikrant8051 commented:
In process of solving previous mentioned issue. Try to solve it today itself.

vikrant8052 committed:
samples: mesh: nrf52: improvement in status responses details

If transition is in process, then as per current implementation bound
states only respond with Present value. But with this commit, now
bound states too would respond with present, target & remaining time
values in status response.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
Mentioned issue has solved now completely !!
Please review someone.

vikrant8051 commented:
@jhedberg Please review latest changes.

For previously mentioned issue (regarding details in status response even when bound state is in transition) I've to modify architecture aggressively (last commit).

vikrant8052 committed:
samples: mesh: nrf52: removed unnecessary Models

Generic OnOff Server & Client has removed from Second
element which is redundant.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c:
```diff
@@ -0,0 +1,166 @@
+#include "device_composition.h"
+#include "storage.h"
+
+static u8_t STORAGE_ID;
```
Only macros should be capital case

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c:
```diff
@@ -0,0 +1,166 @@
+{
+	char buf[4];
+
+	settings_str_from_value(SETTINGS_INT8, &reset_counter, buf,
```
Please avoid using any other value types than the "bytes". INT8 etc will be removed in the future. So use settings_str_from_bytes() and settings_bytes_from_str() helpers instead.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c:
```diff
@@ -0,0 +1,166 @@
+	settings_str_from_value(SETTINGS_INT8, &reset_counter, buf,
+				sizeof(buf));
+
+	settings_save_one("ps/1", buf);
```
You might want to use something more friendlier than numbers for the paths, e.g. "ps/rc" (for reset counter). Why "ps" btw? If that's intended to be short for "persistent storage" then it seems a bit unnecessarily generic. Wouldn't it be better to use something specific to the data that's stored under it, e.g. "onoff" or something like that?

vikrant8051 replied:
Here I've change naming style from number to words.
Keep "ps" as it is.

vikrant8051 commented:
@jhedberg 
[storage.txt](https://github.com/zephyrproject-rtos/zephyr/files/2495174/storage.txt)

Kindly go thrugh with attached file. After these changes not able to save Values of SoC flash.

vikrant8051 commented:
@jhedberg Now able to save data on SoC flash after making some additional changes.

vikrant8051 commented:
@MariuszSkamra 
Please execute PTS against this PR.

@jhedberg If other things are okay for you then please merged this PR. It is very heavy now.

vikrant8051 commented:
@carlescufi 
Not able to see progree on Shippable. Also tried to close & re-open this PR but issue hasn't resolved.

dbkinder reviewed (approved):
+1 for doc changes

vikrant8051 commented:
@jhedberg Kindly merge this PR. If anything is pending then I will take of it in next PR.

Hope someone will execute PTS against it before release of v1.14 :)

vikrant8052 committed:
samples: mesh: nrf52: follow the standard while naming variables

Change variable names to follow standard.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
@jhedberg Thanks for merging.

jhedberg commented:
> @jhedberg Thanks for merging this PR.

@vikrant8051 you're welcome. I hope we can avoid your future PRs growing this big - this one really became too big and went on for too long time to do a proper review of it.
